<a href="https://colab.research.google.com/github/ShilpaVinayaka/Machine-learning-projects/blob/main/%20NextWordPrediction/RNNAlgorithm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install keras.utils.to_categorical

ERROR: Could not find a version that satisfies the requirement keras.utils.to_categorical (from versions: none)
ERROR: No matching distribution found for keras.utils.to_categorical


In [ ]:
from numpy import array
import numpy as np
import pandas as pd
from keras.preprocessing.text import Tokenizer
from tensorflow.keras.utils import to_categorical
from keras.models import Sequential
from keras.layers import Dense, LSTM, Embedding

In [ ]:
#loading the dataset to a pandas dataframe
data = pd.read_csv("Sentences.csv")

In [ ]:
data.head()

Time Behaviour Active_state Position Sentences
0  00:00    normal       asleep  inhouse       NaN
1  00:10    normal       asleep  inhouse       NaN
2  00:20    normal       asleep  inhouse       NaN
3  00:30    normal       asleep  inhouse       NaN
4  00:40    normal       asleep  inhouse       NaN

In [ ]:
data.isnull().sum()

Time             0
Behaviour        0
Active_state     0
Position         0
Sentences       79
dtype: int64

In [ ]:
data = data.fillna('')

In [ ]:
tokenizer = Tokenizer()

In [ ]:
tokenizer.fit_on_texts(data['Sentences'].values)

In [ ]:
encoded_Data = tokenizer.texts_to_sequences(data['Sentences'])

In [ ]:
encoded_Data

[[],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [5,
  8,
  1,
  17,
  50,
  17,
  40,
  51,
  18,
  73,
  9,
  52,
  19,
  132,
  133,
  18,
  73,
  9,
  52,
  19,
  134],
 [5,
  8,
  1,
  17,
  50,
  17,
  40,
  51,
  1,
  8,
  20,
  1,
  6,
  3,
  41,
  42,
  1,
  32,
  36,
  7,
  43,
  1,
  6,
  3,
  36,
  7,
  43,
  1,
  6,
  3,
  74,
  7,
  75,
  1,
  6,
  3,
  24,
  1,
  6,
  19,
  76,
  77,
  11],
 [],
 [5,
  8,
  1,
  17,
  50,
  17,
  40,
  51,
  1,
  8,
  20,
  1,
  6,
  3,
  41,
  42,
  1,
  32,
  36,
  7,
  43,
  1,
  6,
  3,
  36,
  7,
  43,
  1,
  6,
  3,
  74,
  7,
  75,
  1,
  6,
  3,
  24,
  5,
  2,
  15,
  135,
  36,
  136,
  19,
  137,
  44],
 [33, 44, 25, 12, 4, 53, 1, 6, 3, 10, 54, 1, 13, 10, 14, 55],
 [33,
  44,
  25,
  12,
  4,
  53,
  1,
  6,
  3,
  10,
  54,
  1,
  13,
  10,
  14,
  55,
  5,
  2,
  78,
  1,
  8,
  20,
  1,
  21,
  11,
  1,
  13,
  56,
  11,
  1,
  13,
  79,


In [ ]:
vocab_size = len(tokenizer.word_index) + 1
print("Vocabulary Size: %d" % vocab_size)

Vocabulary Size: 201


We need to create sequences of words to fit the model with one word as input and one word as output

In [ ]:
# create word -> word sequences
sequences = list()
for i in range(1, len(encoded_Data)):
  for j in range(len(encoded_Data[i])):
    sequence = encoded_Data[i][j-1:j+1]
    sequences.append(sequence)
print("Total sequences: %d" % len(sequences))

Total sequences: 884


Running this peice shows that we have a total of 884 input-output pairs to train the network

In [ ]:
sequences
# [input, output]

[[],
 [5, 8],
 [8, 1],
 [1, 17],
 [17, 50],
 [50, 17],
 [17, 40],
 [40, 51],
 [51, 18],
 [18, 73],
 [73, 9],
 [9, 52],
 [52, 19],
 [19, 132],
 [132, 133],
 [133, 18],
 [18, 73],
 [73, 9],
 [9, 52],
 [52, 19],
 [19, 134],
 [],
 [5, 8],
 [8, 1],
 [1, 17],
 [17, 50],
 [50, 17],
 [17, 40],
 [40, 51],
 [51, 1],
 [1, 8],
 [8, 20],
 [20, 1],
 [1, 6],
 [6, 3],
 [3, 41],
 [41, 42],
 [42, 1],
 [1, 32],
 [32, 36],
 [36, 7],
 [7, 43],
 [43, 1],
 [1, 6],
 [6, 3],
 [3, 36],
 [36, 7],
 [7, 43],
 [43, 1],
 [1, 6],
 [6, 3],
 [3, 74],
 [74, 7],
 [7, 75],
 [75, 1],
 [1, 6],
 [6, 3],
 [3, 24],
 [24, 1],
 [1, 6],
 [6, 19],
 [19, 76],
 [76, 77],
 [77, 11],
 [],
 [5, 8],
 [8, 1],
 [1, 17],
 [17, 50],
 [50, 17],
 [17, 40],
 [40, 51],
 [51, 1],
 [1, 8],
 [8, 20],
 [20, 1],
 [1, 6],
 [6, 3],
 [3, 41],
 [41, 42],
 [42, 1],
 [1, 32],
 [32, 36],
 [36, 7],
 [7, 43],
 [43, 1],
 [1, 6],
 [6, 3],
 [3, 36],
 [36, 7],
 [7, 43],
 [43, 1],
 [1, 6],
 [6, 3],
 [3, 74],
 [74, 7],
 [7, 75],
 [75, 1],
 [1, 6],
 [6, 3],
 [3, 24

We can then split the sequences into input (X) and output elements (Y)

In [ ]:
sequences = [x for x in sequences if x]
sequences

[[5, 8],
 [8, 1],
 [1, 17],
 [17, 50],
 [50, 17],
 [17, 40],
 [40, 51],
 [51, 18],
 [18, 73],
 [73, 9],
 [9, 52],
 [52, 19],
 [19, 132],
 [132, 133],
 [133, 18],
 [18, 73],
 [73, 9],
 [9, 52],
 [52, 19],
 [19, 134],
 [5, 8],
 [8, 1],
 [1, 17],
 [17, 50],
 [50, 17],
 [17, 40],
 [40, 51],
 [51, 1],
 [1, 8],
 [8, 20],
 [20, 1],
 [1, 6],
 [6, 3],
 [3, 41],
 [41, 42],
 [42, 1],
 [1, 32],
 [32, 36],
 [36, 7],
 [7, 43],
 [43, 1],
 [1, 6],
 [6, 3],
 [3, 36],
 [36, 7],
 [7, 43],
 [43, 1],
 [1, 6],
 [6, 3],
 [3, 74],
 [74, 7],
 [7, 75],
 [75, 1],
 [1, 6],
 [6, 3],
 [3, 24],
 [24, 1],
 [1, 6],
 [6, 19],
 [19, 76],
 [76, 77],
 [77, 11],
 [5, 8],
 [8, 1],
 [1, 17],
 [17, 50],
 [50, 17],
 [17, 40],
 [40, 51],
 [51, 1],
 [1, 8],
 [8, 20],
 [20, 1],
 [1, 6],
 [6, 3],
 [3, 41],
 [41, 42],
 [42, 1],
 [1, 32],
 [32, 36],
 [36, 7],
 [7, 43],
 [43, 1],
 [1, 6],
 [6, 3],
 [3, 36],
 [36, 7],
 [7, 43],
 [43, 1],
 [1, 6],
 [6, 3],
 [3, 74],
 [74, 7],
 [7, 75],
 [75, 1],
 [1, 6],
 [6, 3],
 [3, 24],
 [24, 5],
 [

In [ ]:
len(sequences)

819

In [ ]:
sequences = array(sequences)
sequences

array([[  5,   8],
       [  8,   1],
       [  1,  17],
       ...,
       [ 63, 123],
       [123,  70],
       [ 70,  48]])

In [ ]:
sequences[0]

array([5, 8])

In [ ]:
sequences[0][0]

5

In [ ]:
sequences = array(sequences)
X, Y = [], []

for i in range(len(sequences)):
  a = sequences[i][0]
  X.append(a)
  b = sequences[i][1]
  Y.append(b)


In [ ]:
X[:5]

[5, 8, 1, 17, 50]

In [ ]:
Y[:5]

[8, 1, 17, 50, 17]

In [ ]:
# one hot encode outputs
Y = to_categorical(Y, num_classes = vocab_size)
# define model
Y[:5]

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

In [ ]:
model = Sequential()
model.add(Embedding(vocab_size, 10, input_length=1))
model.add(LSTM(50))
model.add(Dense(vocab_size, activation='softmax'))
print(model)

In [ ]:
model

In [ ]:
#compile network
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
#fit network
X = np.asarray(X)
Y = np.asarray(Y)
model.fit(X, Y, epochs=100)

In [ ]:
#generate a sequence from the model
def generate_seq(model, tokenizer, enter_text, n_pred):
  in_text, result = enter_text, enter_text
  # generate fixed number of words
  for _ in range(n_pred):
    #encode the text as integer
    encoded = tokenizer.texts_to_sequences([in_text])[0]
    encoded = array(encoded)
  
    # predict a word in the vocalbulary
    yhat = model.predict(encoded)
    
    # map predicted word index to word
    out_word = ''
    for word,index in tokenizer.word_index.items():
      
      if index == yhat.all():
        out_word = word
        break

    # append to input
    in_text, result = out_word, result + ' ' + out_word
  return result

In [ ]:
print(generate_seq(model, tokenizer, "Is it time for", 2))

Is it time for i i
